File: implement-shell-tools/ls/ls.js
import { program } from "commander";
import { promises as fs } from "node:fs";

program
  .name("ls")
  .description("List directory contents")
  .option("-a, --all", "show hidden files")
  .option("-1, --one-per-line", "display one file per line")
  .argument("[path]", "directory to list", ".");

program.parse();

const options = program.opts();
const path = program.args[0] || ".";

const files = await fs.readdir(path);

const filteredFiles = files.filter((file) => {
  return options.all || !file.startsWith(".");
});

if (options.onePerLine) {
  for (const file of filteredFiles) {
    console.log(file);
  }
} else {
  console.log(filteredFiles.join(" "));
}

File: implement-shell-tools/ls/ls.py
import argparse
import os

parser = argparse.ArgumentParser()
parser.add_argument("-a", "--all", action="store_true")
parser.add_argument("-1", dest="one_per_line", action="store_true")
parser.add_argument("path", nargs="?", default=".")

args = parser.parse_args()

files = sorted(os.listdir(args.path))

files = [f for f in files if args.all or not f.startswith(".")]

if args.one_per_line:
    for f in files:
        print(f)
else:
    print(" ".join(files))